### ISSUE TITLE
Details

### ISSUE THREAD

**Issue by arita37** [deezer/gravity_graph_autoencoders#9]
Hello,

   Just one quick question on this,
      does the GI  AE  algo can scale to 10 mio of items,  100 mio of items ?
      
   One common issue is to have things that scale, just outside of a local laptop compute.

**Comment by GuillaumeSalhaGalvan** [comment]
Dear @arita37,

Thank you for your message.

We discuss scalability concerns in Section 3.6 of [our paper](https://arxiv.org/pdf/1905.09570.pdf). In essence, the code provided in this repository suffers from a quadratic complexity w.r.t. the number of nodes _n_ in the graph. At each training iteration, we need to reconstruct entire _n x n_ adjacency matrices to evaluate reconstruction losses or ELBO objectives. As a consequence, on your own laptop, you will likely get errors when trying to reconstruct graphs with (roughly) _n > 40000_ or _50000_ nodes. Note: this issue is not specific to gravity-inspired models, e.g., you would encounter similar errors when running standard GAE and VGAE models using [T.Kipf's original code](https://github.com/tkipf/gae), as it also reconstructs entire adjacency matrices.

Do you need to train standard or gravity-inspired GAE/VGAE models on larger graphs?
- Firstly, I would advice against adopting "naive" uniform mini-batch sampling strategies. By experience, this would often significantly deteriorate performances (intuitively, if you randomly reconstruct 1% of nodes/edges drawn at random, you do not necessarily get a good summary of the initial graph structure);
- However, there exists more effective yet simple sampling strategies to scale GAE/VGAE to large graphs with millions of nodes and edges. You can take a look at [our recent paper introducing the FastGAE method](https://arxiv.org/pdf/2002.01910.pdf).

I am currently finishing my Ph.D. thesis. After my defense (~Feb. 2022), I want to incorporate FastGAE into this repository to improve scalability. Meanwhile, you can check the work of @fleverest who recently combined these two methods in [one of his own repositories](https://github.com/fleverest/Gravity_FastGAE).

Best regards,

Guillaume

**Comment by arita37** [comment]
Thanks for the details.

Usually, have 40 million items (nodes),
scaling is 1st priority, 
otherwise cannot be tested meanfully
(toy dataset does not mean much…)


Besides,
do you think it can be integrated into
existing distributed GNN framework ?

https://docs.dgl.ai/performance.html

Thanks

**Comment by GuillaumeSalhaGalvan** [comment]
I totally understand! At Deezer, most of our graphs also have millions of nodes, e.g., music artists or users.

Honestly, I am not very familiar with DGL. I didn't see any complete implementation of GAE/VGAE (maybe I missed it). However, the library seems to already contain all building blocks (including GCN encoder, inner product or embedding distance computations and cross entropy losses)  to integrate these models.

**Comment by arita37** [comment]
Sure, thanks
Yes, main part is usually on the scalability part,
especially for graph structure…

Is the updated version 
quadratic in number of nodes ?


Here, one implementation of GAE in DGL:
https://github.com/shionhonda/gae-dgl/blob/master/gae_dgl/gae.py

**Comment by GuillaumeSalhaGalvan** [comment]
The complexity of FastGAE depends on the sizes of stochastic subgraphs drawn for decoding, which is a selectable parameter. For specific sizes, motivated in [the paper](https://arxiv.org/pdf/2002.01910.pdf), the complexity evolves linearly, while still reaching fairly good performances on link prediction and community detection tasks (see, e.g., Section 3.4 and experiments for more details).

**Comment by GuillaumeSalhaGalvan** [comment]
I'm closing this issue, but feel free to contact me if needed!

**Comment by arita37** [comment]
Thanks for the various feedback.

For the info,
Deepmind releases a paper
about constrating loss
without negative sampling,
which scales very well.
